### ISSUE TITLE
training problem

### ISSUE THREAD

**Issue by lianghao2000** [sanghviyashiitb/poisson-deblurring#3]
Thank you very much for the training code provided by the author.  

I use trained model and your pre-trained model to test the Levin dataset, but there are some issues.  The objective metric (**PSNR**) as follows.

1. pre-trained model provided by you: 20.056
2. trained model by your training codes: 20.1841
3. trained model by my training codes: 20.629

But, in another of your papers named "Photon-Limited Blind Deconvolution using Unsupervised Iterative Kernel Estimation", The PSNR result of P4IP method t is 22.45 under the same conditions and data set.
![](https://qiniu.lianghao.work/20221206110340.png).

***I would like to ask which part I need to improve in order to achieve the best results, or where my problems arise. Please correct me. Good luck.***

Note: the **Poisson peak** is **10**, and test dataset is Levin, you can get it from [Google Drive](https://drive.google.com/drive/folders/1xhewhENk7J8VymdOyyU11JkzOAghsyvJ?usp=share_link)

the test codes as follows:
```python
import torch
import numpy as np
from scipy.io import loadmat
from skimage.metrics import peak_signal_noise_ratio as cal_psnr
from models.network_p4ip import P4IP_Net
np.random.seed(20000320)

data_dir = "E:/Project/Image_Processing/Poisson_Deblurring/TestDataSet/Levin.mat"

data = loadmat(data_dir)["data"]

psnr = 0.0
peak = 10.0
model = P4IP_Net()
# model_file = "docs/p4ip_100epoch.pth" # pre-trained
model_file = "model_zoo/p4ip_net_6epoch.pth"
checkpoint = torch.load(model_file)
model.load_state_dict(checkpoint)
model = model.cuda()

for i in range(data.shape[0]):
    sharp, blurry, kernel = data[i]
    blurry = np.float32(np.random.poisson(blurry * peak))
    kernel = np.float32(kernel)
    peak_m = np.array([peak])
    peak_m = peak_m[np.newaxis, np.newaxis, np.newaxis,]
    kernel = kernel[np.newaxis, np.newaxis, ]
    blurry = blurry[np.newaxis, np.newaxis, ]
    # 转为tensor
    blurry_t = torch.from_numpy(blurry.copy())
    kernel_t = torch.from_numpy(kernel.copy())
    peak_t = torch.tensor(peak_m, dtype=torch.float64)
    blurry_t = blurry_t.cuda()
    kernel_t = kernel_t.cuda()
    peak_t = peak_t.cuda()
    restores = model(blurry_t, kernel_t, peak_t)
    restored = restores[-1]
    restored_m = restored.data.cpu().numpy().squeeze()
    psnr += cal_psnr(image_true=sharp, image_test=restored_m)

print(psnr / data.shape[0])
```

**Comment by sanghviyashiitb** [comment]
Hi [HLImg](https://github.com/HLImg),
One reason I can think of for this gap is that you may not be symmetrically padding the input and cropping out the output before using images from the Levin dataset which contain real-blur. You can use the helper function 'p4ip_wrapper_pad' in the utils_test file which can do that for you. 
To further understand why this modification is required you can refer to Section IV of supplementary document [here](https://arxiv.org/pdf/2110.15314.pdf).

Please let me know if this issue still persists. In that case I will upload the testing script for levin dataset by the end of the week.

**Comment by lianghao2000** [comment]
> Hi [HLImg](https://github.com/HLImg), One reason I can think of for this gap is that you may not be symmetrically padding the input and cropping out the output before using images from the Levin dataset which contain real-blur. You can use the helper function 'p4ip_wrapper_pad' in the utils_test file which can do that for you. To further understand why this modification is required you can refer to Section IV of supplementary document [here](https://arxiv.org/pdf/2110.15314.pdf).
> 
> Please let me know if this issue still persists. In that case I will upload the testing script for levin dataset by the end of the week.

Thanks for your reply. I use the **p4ip_wrapper_pad**  to test, but the metric result is terrible, ***psnr = 10.769009537470783***, I hope your testing scripts for levin dataset. 
My test scripts as follows, I use your pre-trained model to test.
````python
import torch
import numpy as np
from scipy.io import loadmat
from skimage.metrics import peak_signal_noise_ratio as cal_psnr
from models.network_p4ip import P4IP_Net
np.random.seed(20000320)

data_dir = "E:/Project/Image_Processing/Poisson_Deblurring/TestDataSet/Levin.mat"

data = loadmat(data_dir)["data"]

psnr = 0.0
peak = 10.0
model = P4IP_Net()
model_file = "docs/p4ip_100epoch.pth" # pre-trained
# model_file = "model_zoo/p4ip_net_6epoch.pth"
checkpoint = torch.load(model_file)
model.load_state_dict(checkpoint)
model = model.cuda()

from utils.utils_test import p4ip_wrapper_pad

for i in range(data.shape[0]):
    sharp, blurry, kernel = data[i]
    restored = p4ip_wrapper_pad(blurry, kernel, peak, model)
    # restored_m = restored.data.cpu().numpy().squeeze()
    psnr += cal_psnr(image_true=sharp, image_test=restored)

print(psnr / data.shape[0])
```

**Comment by sanghviyashiitb** [comment]
Can you share Levin.mat with me so I can debug the code better?

**Comment by lianghao2000** [comment]
> Can you share Levin.mat with me so I can debug the code better?

https://drive.google.com/drive/folders/1xhewhENk7J8VymdOyyU11JkzOAghsyvJ

**Comment by HLImg** [comment]
Hi, I note the difference in the **PSNR calculation** between what I used and what you used. I use the python package 
```python
from skimage.metrics import structural_similarity as cal_ssim
from skimage.metrics import peak_signal_noise_ratio as cal_psnr
```
you can install it by `pip install scikit-image`

**Comment by sanghviyashiitb** [comment]
Hi HLimg, 

Thank you for your patience, here is the code snippet which works, and gives PSNR: 21.75, SSIM: 0.581

```
import numpy as np
import torch
import torch.nn as nn

from skimage.metrics import structural_similarity as ssim

from PIL import Image
from models.network_p4ip import P4IP_Net
from utils.utils_torch import img_to_tens, scalar_to_tens
import matplotlib.pyplot as plt
from scipy.io import loadmat
np.random.seed(20)


def shift_inv_metrics(x1, x2, search_window=20):
	x1 = np.clip(x1, 0, 1)
	max_psnr = -np.inf
	max_ssim = -np.inf
	for i in np.arange(-search_window, search_window+1):
		for j in np.arange(-search_window, search_window+1):
			x0 = np.roll(x1, (i,j), axis=[0,1])
			mse = np.mean((x0-x2)**2)
			psnr =  -10*np.log10(mse)
			ssim_val = ssim(x2, x0, data_range=np.max(np.ravel(x0))-np.min(np.ravel(x0)))

			if psnr > max_psnr:
				max_psnr = psnr
			if ssim_val > max_ssim:
				max_ssim = ssim_val

	return np.asarray([max_psnr, max_ssim])


def p4ip_wrapper(y, k, M, p4ip, mode ='circular'):
	device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
	if mode == 'symmetric':
		H, W = np.shape(y)
		H1, W1 = H//4, W//4
		y = np.pad(y, ((H1,H1),(W1,W1)), mode='symmetric')
	
	Ht = img_to_tens(k).to(device).float()	
	yt = img_to_tens(y).to(device)
	Mt = scalar_to_tens(M).to(device)
	with torch.no_grad():
		x_rec_list = p4ip(yt, Ht, Mt)
	x_rec = x_rec_list[-1]
	x_rec = np.clip(x_rec.cpu().detach().numpy(),0,1)	
	x_out = x_rec[0,0,:,:]
	if mode == 'symmetric':
		x_out = x_out[H1:H+H1, W1:W+W1]
	return x_out




LEVIN_DIR = '../datasets/levin/groundtruth/'
ALPHA = np.float32(10.0)

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
nb_solver = P4IP_Net()
nb_solver.load_state_dict(torch.load('model_zoo/p4ip_latest.pth'))
nb_solver.to(device); nb_solver.eval()

data = loadmat('data/Levin.mat')["data"]
metric_arr = np.zeros([2,data.shape[0]], dtype=np.float32)
for i in range(data.shape[0]):	
	x, yn, kernel = data[i]
	y = np.random.poisson(np.maximum(ALPHA*yn,0)).astype(np.float32)	
	with torch.no_grad():
		x_nb = p4ip_wrapper(y, kernel, ALPHA, nb_solver, 'symmetric')
	metric_arr[:,i] = shift_inv_metrics(x_nb, x)

avg_metrics = metric_arr.mean(axis=(1))
print('PSNR: %0.2f, SSIM: %0.3f'%(avg_metrics[0], avg_metrics[1]))
```
There are a few things needed to get to 22.45dB PSNR from your code.
1. Firstly, please note some minor tweaks were made to the model architecture for the blind deconvolution paper you referred to in the original post. These changes are now updated in this repository itself and the corresponding trained model is also updated with link [here](https://drive.google.com/file/d/1SimYIMDOijOBV_MTepM9k6xAZQHWIafo/view?usp=share_link).

2. Since blind deconvolution algorithms are subject to shift of the image, we use a helper function called 'shift_inv_metrics' which moves the image around for the best PSNR and SSIM. We use that for all the algorithms in Table 1 of the paper, including the non-blind solver.

3. Most importantly, our result is still ~0.7dB ahead of this script. This is very likely because we use the .png files as the input which may introduce quantization effects and hence the improved  PSNR.

I intend to upload the code for the blind deconvolution paper this January as I am going on a vacation for the next 3 weeks. I will make sure to include the evaluation script for Levin dataset when that code is made public.

**Comment by lianghao2000** [comment]
Thank you for your patience. Have a nice vocation.

**Comment by sanghviyashiitb** [comment]
The code for blind deconvolution is available here : https://github.com/sanghviyashiitb/photon-limited-blind
